core/include/vecmem/containers/device_array.ipp

```diff
@@ -0,0 +1,271 @@
+/** VecMem project, part of the ACTS project (R&D line)
+ *
+ * (c) 2021 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+#pragma once
+
+// System include(s).
+#include <cassert>
+
+namespace vecmem {
+
+   template< typename T, std::size_t N >
+   VECMEM_HOST_AND_DEVICE
+   device_array< T, N >::device_array( device_vector_data< value_type > data )
+   : m_ptr( data.m_ptr ) {
+
+      assert( data.m_size >= N );
+   }
+
+   template< typename T, std::size_t N >
+   VECMEM_HOST_AND_DEVICE
+   device_array< T, N >::device_array( const device_array& parent )
+   : m_ptr( parent.m_ptr ) {
+
+   }
+
+   template< typename T, std::size_t N >
+   VECMEM_HOST_AND_DEVICE
+   device_array< T, N >&
+   device_array< T, N >::operator=( const device_array& rhs ) {
+
+      // Prevent self-assignment.
+      if( this == &rhs ) {
+         return *this;
+      }
+
+      // Copy the other object's payload.
+      m_ptr = rhs.m_ptr;
+
+      // Return a reference to this object.
+      return *this;
+   }
+
+   template< typename T, std::size_t N >
+   VECMEM_HOST_AND_DEVICE
+   typename device_array< T, N >::reference
+   device_array< T, N >::at( size_type pos ) {
+
+      // Check if the index is valid.
+      assert( pos < N );
+
+      // Return a reference to the vector element.
+      return m_ptr[ pos ];
+   }
+
+   template< typename T, std::size_t N >
+   VECMEM_HOST_AND_DEVICE
+   typename device_array< T, N >::const_reference
+   device_array< T, N >::at( size_type pos ) const {
+
+      // Check if the index is valid.
+      assert( pos < N );
+
+      // Return a reference to the vector element.
+      return m_ptr[ pos ];
+   }
+
+   template< typename T, std::size_t N >
+   VECMEM_HOST_AND_DEVICE
+   typename device_array< T, N >::reference
+   device_array< T, N >::operator[]( size_type pos ) {
+
+      // Return a reference to the vector element.
+      return m_ptr[ pos ];
+   }
+
+   template< typename T, std::size_t N >
+   VECMEM_HOST_AND_DEVICE
+   typename device_array< T, N >::const_reference
+   device_array< T, N >::operator[]( size_type pos ) const {
+
+      // Return a reference to the vector element.
+      return m_ptr[ pos ];
+   }
+
+   template< typename T, std::size_t N >
+   VECMEM_HOST_AND_DEVICE
+   typename device_array< T, N >::reference
+   device_array< T, N >::front() {
+
+      // Make sure that there is at least one element in the vector.
+      static_assert( N > 0, "Cannot return first element of empty array" );
```

Comment:
Should we put these static asserts in the class instead of in its methods? Creating an array of length 0 makes little sense at the type level.

Reply:
Possibly... Though every other functionality of the code works fine with a zero sized container. And I could imagine using some type of template meta-programming that could result in instantiating zero sized arrays.